databricks/koalas/indexes.py

```diff
@@ -1631,6 +1632,67 @@ def repeat(self, repeats: int) -> "Index":
         else:
             return ks.concat([kdf] * repeats).index
 
+    def union(self, other, sort=None):
+        """
+        Form the union of two Index objects.
+
+        Parameters
+        ----------
+        other : Index or array-like
+        sort : bool or None, default None
+            Whether to sort the resulting Index.
+
+        Returns
+        -------
+        union : Index
+
+        Examples
+        --------
+
+        Index
+
+        >>> idx1 = ks.Index([1, 2, 3, 4])
+        >>> idx2 = ks.Index([3, 4, 5, 6])
+        >>> idx1.union(idx2).sort_values()
+        Int64Index([1, 2, 3, 4, 5, 6], dtype='int64')
+
+        MultiIndex
+
+        >>> midx1 = ks.MultiIndex.from_tuples([("x", "a"), ("x", "b"), ("x", "c"), ("x", "d")])
+        >>> midx2 = ks.MultiIndex.from_tuples([("x", "c"), ("x", "d"), ("x", "e"), ("x", "f")])
+        >>> midx1.union(midx2).sort_values()  # doctest: +SKIP
+        MultiIndex([('x', 'a'),
+                    ('x', 'b'),
+                    ('x', 'c'),
+                    ('x', 'd'),
+                    ('x', 'e'),
+                    ('x', 'f')],
+                   )
+        """
+        sort = True if sort is None else sort
+        sort = validate_bool_kwarg(sort, "sort")
+        if not isinstance(other, Index) and isinstance(self, MultiIndex):
+            if isinstance(other, list) and not all([isinstance(item, tuple) for item in other]):
+                raise TypeError("other must be a MultiIndex or a list of tuples")
+            other = MultiIndex.from_tuples(other)
+        if not isinstance(other, Index) and not isinstance(self, MultiIndex):
+            other = Index(other)
+        if type(self) is not type(other):
+            # TODO: We can't support different type of values in a single column for now.
+            raise NotImplementedError("Union between Index and MultiIndex is not yet supported")
+        sdf_self = self._internal._sdf.select(self._internal.index_spark_columns)
+        sdf_other = other._internal._sdf.select(
+            other._internal.index_spark_columns
+        ).drop_duplicates()
+        sdf = sdf_self.union(sdf_other.subtract(sdf_self))
```

Comment:
Do we need such a complex plan? when?

Reply:
Let me show you some examples of `union` in pandas.

If there are duplicated values ​​in `other` which among the common values between `self` and `other`, pandas just doesn't use those values. So I use `sdf_other.subtract(sdf_self)` first.

```python
>>> idx1 = pd.Index([1, 2, 3, 4])
>>> idx2 = pd.Index([3, 4, 3, 4, 3, 4, 5, 6])
>>> idx1.union(idx2)
Int64Index([1, 2, 3, 4, 5, 6], dtype='int64')
```

But if there are duplicated values in `self` which among the common value between `self` and `other`, pandas use them all. So I just union them. (self and subtracted other)

```python
>>> idx1 = pd.Index([1, 2, 3, 4, 3, 4, 3, 4])
>>> idx2 = pd.Index([3, 4, 5, 6])
>>> idx1.union(idx2)
Int64Index([1, 2, 3, 3, 3, 4, 4, 4, 5, 6], dtype='int64')
```

Reply:
Thanks for the examples. Seems like it's a behavior change in pandas>=1.0.
Shall we add the example in the tests?

btw, I guess we don't need `drop_duplicates()` in line 1686:

```py
>>> idx1
Int64Index([1, 2, 3, 4, 3, 4, 3, 4], dtype='int64')
>>> idx2
Int64Index([3, 4, 3, 4, 3, 4, 3, 4, 5, 6, 5, 6, 5, 6], dtype='int64')
>>> idx1.union(idx2)
Int64Index([1, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 6], dtype='int64')
```

Reply:
@ueshin That's right. The behavior has changed since pandas>=1.0.

Okay, then I'm gonna add the tests again and make them run when only pandas>=1.0 .
(Actually, I removed them in https://github.com/databricks/koalas/pull/1348/commits/16a0a44f6ec2c8b4765535568b8c644a7b1b19c1 to compatibility with pandas<1.0 at the first time 😹)

Reply:
> btw, I guess we don't need `drop_duplicates()` in line 1686:

Oh, It looks so, I'll fix them. Thanks for the comment!